src/soter/openssl/soter_kdf.c

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2019 Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "soter/soter_kdf.h"
+
+#include <limits.h>
+
+#include <openssl/evp.h>
+
+soter_status_t soter_pbkdf2_sha256(const uint8_t* passphrase,
+                                   size_t passphrase_length,
+                                   const uint8_t* salt,
+                                   size_t salt_length,
+                                   size_t iterations,
+                                   uint8_t* key,
+                                   size_t key_length)
+{
+    int res;
+
+    SOTER_CHECK_PARAM(passphrase != NULL);
+    SOTER_CHECK_PARAM(passphrase_length > 0);
+    SOTER_CHECK_PARAM(passphrase_length <= INT_MAX);
+    if (!salt) {
+        SOTER_CHECK_PARAM(salt_length == 0);
+    }
+    SOTER_CHECK_PARAM(salt_length <= INT_MAX);
+    SOTER_CHECK_PARAM(iterations > 0);
+    SOTER_CHECK_PARAM(iterations <= INT_MAX);
+    SOTER_CHECK_PARAM(key != NULL);
```

Comment:
The implementation here does not follow our standard pattern of getting the output size: not everyone knows that `sha256` is 32 bytes and how does `hmac-sha256` contribute to it.

Reply:
> The implementation here does not follow our standard pattern of getting the output size

That’s because there are no specific requirements for the output size as there are with other APIs. It can be any positive number, PBKDF2 will fill the buffer. Typical users of this API know exactly how many key bytes they need for the algorithm they use. If Secure Cell uses AES-128 then it will request just 16 bytes.

We could provide the usual API which outputs recommended output length when `key == NULL`, and when given zero output size will return SOTER_BUFFER_TOO_SMALL rather than SOTER_INVALID_PARAM. However, no API in Themis will use this code path, and probably none of the advanced users that need this function will use it too.

Other APIs output expected buffer size because they have specific requirements for that and will not work if they are not fulfilled. For encryption/decryption the output buffer must have enough space to fit the output (and the metadata in case of Themis). Generated asymmetric keys must have specific size and use custom format. On the other hand, symmetric keys may have arbitrary size and are just blobs, yet Themis API provides size suggestion. This is because *those* users are not typically aware of the recommended key size (but may still use any size they want). Consider `soter_rand()` which does not provide suggestions about the recommended size of the random output, it just fills whatever buffer it has been given. Similarly to that, key derivation does not provide suggestions because the caller of this advanced API will know better. This would be a different story if the output contained some additional metadata or padding. For example, encryption/decryption functions use output parameters for output size because the output may have padding added or removed. This is not the case with key derivation where we get exactly as many bytes as we request.

To sum it up, I think that providing API for output key length suggestion is unnecessary complication which is better served by documentation. But I’m not opposed to it as such.

Any second opinion?

Reply:
Well, if we're not exposing this API outside via Themis, then it is not that required. However, I would put it here, just for future usage. If the users will just go through public Themis API, implementation of which already "knows" the length of the buffer needed, then, most likely this code will be a no-op anyway.